drud/ddev
# Add ddev-ssh-auth container to provide ssh-auth, fixes #414, fixes #1152

## The Problem/Issue/Bug:

People want to be able to use their ssh keys inside the web container, mostly for composer actions against private repositories, but also for things like drush access and direct ssh access.

## How this PR Solves The Problem:

* Adds a ddev-ssh-auth container that serves all other ddev projects with ssh-auth running
* Incidentally solves #1152 (request to allow turning off phpMyAdmin container) with an omit-containers array in config.yaml, which was added in order to allow turning off the ddev-ssh-agent container (but might as well turn off dba=phpMyAdmin as well)

## Manual Testing Instructions:

This will probably need a special ddev command for step 2, but for now, do this:

1. `ddev start` a project
2. Assuming your password-protected ssh key is in ~/.ssh/, run `ddev auth ssh` and give it your key password. 
3. In a project, `ddev ssh`. You should be able to `ssh <username>@somessh-protected-host` successfully. If you don't have something useful to test with, give me your pubkey and I'll set you up.
4. Put a password-protected ssh key in another directory and use `ddev auth ssh -d /somewhere/else` to add it (and test that you can now access what you ought to be able to access)
5. Try using `ddev rm --stop-ssh-agent` to stop the ssh agent container; when you do that, auth from within the web container should be lost; to re-establish it a `ddev start` should get it going again, and ddev auth ssh is required again.
6. Try omit_containers: ["ddev-ssh-agent", "dba"] with a `ddev start` to verify that the dba (phpMyAdmin) and ddev-ssh-agent containers don't start.

## Automated Testing Overview:

This doesn't yet have tests and it will need some.

## Related Issue Link(s):

OP #414 
Random issue about omitting containers #1152 

## Release/Deployment notes:

## TODO: 

- [x] Needs tests
- [x] The shared volume may need to be deleted at first run of the ddev-ssh-agent (right before running it?)
- [x] ddev probably needs a special command to add the ssh key. `ddev auth ssh` ?
- [x] Figure out the correct way for users to kill off the ddev-ssh-agent if they don't want it
- [x] Consider configuration for disabling the ddev-ssh-agent
- [x] Add ddev-ssh-agent status to ddev describe (maybe ddev list?)
- [x] Add ddev-ssh-agent version to `ddev version`
- [x] Consider whether we need a command-line switch to `ddev config` for omit_containers.
